leveldb/db_compaction.go

```diff
@@ -786,80 +850,324 @@ func (db *DB) mCompaction() {
+		// db operation(e.g. writes).
+		if len(waitQ) > 0 && db.resumeWrite() {
 			for i := range waitQ {
 				waitQ[i].ack(nil)
 				waitQ[i] = nil
 			}
 			waitQ = waitQ[:0]
+		}
+		var (
+			needCompact bool
+			level       int
+			table       *tFile
+		)
+		if ctx.count() < compLimit && rangeCmd == nil {
+			needCompact, level, table = db.tableNeedCompaction(ctx)
+		}
+		if needCompact {
 			select {
+			case <-db.closeC:
+				return
 			case x = <-db.tcompCmdC:
-			case ch := <-db.tcompPauseC:
-				db.pauseCompaction(ch)
+			//case ch := <-db.tcompPauseC:
+			//	db.pauseCompaction(ch)
+			//	continue
+			case c := <-done:
+				ctx.delete(c)
 				continue
+			default:
+			}
+		} else {
+			// If there is a pending range compaction, do it right now
+			if rangeCmd != nil && ctx.count() == 0 {
+				cmd := rangeCmd.(cRange)
+				cmd.ack(db.tableRangeCompaction(cmd.level, cmd.min, cmd.max))
+				rangeCmd = nil
+				continue // Re-loop is necessary, try to spin up more compactions
+			}
+			// If the waitAll list is not empty, send the ack if all compactions have finished.
+			if len(waitAll) != 0 && ctx.count() == 0 {
+				for _, wait := range waitAll {
+					wait.ack(nil)
+					wait = nil
+				}
+				waitAll = waitAll[:0]
+			}
+			select {
 			case <-db.closeC:
 				return
+			case x = <-db.tcompCmdC:
+			//case ch := <-db.tcompPauseC:
+			//	db.pauseCompaction(ch)
+			//	continue
+			case c := <-done:
+				ctx.delete(c)
+				continue
 			}
 		}
 		if x != nil {
 			switch cmd := x.(type) {
 			case cAuto:
-				if cmd.ackC != nil {
-					// Check the write pause state before caching it.
-					if db.resumeWrite() {
-						x.ack(nil)
-					} else {
-						waitQ = append(waitQ, x)
-					}
+				if cmd.full {
+					waitAll = append(waitAll, cmd)
+				} else if cmd.ackC != nil {
+					waitQ = append(waitQ, x)
 				}
 			case cRange:
-				x.ack(db.tableRangeCompaction(cmd.level, cmd.min, cmd.max))
+				if ctx.count() > 0 {
+					rangeCmd = x
+				} else {
+					x.ack(db.tableRangeCompaction(cmd.level, cmd.min, cmd.max))
+				}
 			default:
 				panic("leveldb: unknown command")
 			}
 			x = nil
+			continue
 		}
-		db.tableAutoCompaction()
+		db.runCompaction(ctx, level, table, &subWg, done)
 	}
 }
+
+func (db *DB) runCompaction(ctx *compactionContext, level int, table *tFile, wg *sync.WaitGroup, done chan *compaction) {
+	var c *compaction
+	if table == nil {
+		c = db.s.pickCompactionByLevel(level, ctx)
+		if c == nil {
+			// We can't pick one more isolated compaction in level n.
+			// Mark the entire level as unavailable. In theory it shouldn't
+			// happen a lot.
+			ctx.denylist[level] = struct{}{}
+			return
+		}
+	} else {
+		c = db.s.pickCompactionByTable(level, table, ctx)
+		if c == nil {
+			// The involved tables are not available now. Mark the seek
+			// compaction as unavailable.
+			ctx.noseek = true
+			return
+		}
+	}
+	ctx.add(c)
+	wg.Add(1)
+
+	go func() {
+		// Catch the panic in its own goroutine.
+		defer func() {
```

Comment:
We catch the panic in the goroutine. In the original implementation, the panic will be passed to the main thread. The behaviour is changed...